crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -5,42 +5,71 @@ pub use draw::*;
 pub use draw_state::*;
 
 use bevy_ecs::prelude::{Component, Query};
+use std::cell::Cell;
+use thread_local::ThreadLocal;
 
 use copyless::VecHelper;
 
+pub struct RenderPhaseScope<'a, I: PhaseItem> {
+    phase_queue: &'a mut Vec<I>,
+}
+
+impl<'a, I: PhaseItem> RenderPhaseScope<'a, I> {
+    /// Adds a [`PhaseItem`] to this render phase.
+    #[inline]
+    pub fn add(&mut self, item: I) {
+        self.phase_queue.alloc().init(item);
+    }
+}
+
 /// A resource to collect and sort draw requests for specific [`PhaseItems`](PhaseItem).
 #[derive(Component)]
 pub struct RenderPhase<I: PhaseItem> {
-    pub items: Vec<I>,
+    items: ThreadLocal<Cell<Vec<I>>>,
+    pub sorted: Vec<I>,
 }
 
 impl<I: PhaseItem> Default for RenderPhase<I> {
     fn default() -> Self {
-        Self { items: Vec::new() }
+        Self {
+            items: ThreadLocal::new(),
+            sorted: Vec::new(),
+        }
     }
 }
 
 impl<I: PhaseItem> RenderPhase<I> {
-    /// Adds a [`PhaseItem`] to this render phase.
-    #[inline]
-    pub fn add(&mut self, item: I) {
-        self.items.alloc().init(item);
+    pub fn get(&self) -> &Cell<Vec<I>> {
+        self.items.get_or_default()
+    }
+
+    pub fn phase_scope(&self, f: impl FnOnce(RenderPhaseScope<'_, I>)) {
+        let store = self.get();
+        let mut phase_queue = store.take();
+        f(RenderPhaseScope {
+            phase_queue: &mut phase_queue,
+        });
+        store.set(phase_queue);
     }
 
     /// Sorts all of its [`PhaseItems`](PhaseItem).
     pub fn sort(&mut self) {
-        self.items.sort_by_key(|d| d.sort_key());
+        debug_assert!(self.sorted.is_empty());
+        for batch in self.items.iter_mut() {
+            self.sorted.append(batch.get_mut());
+        }
+        self.sorted.sort_unstable_by_key(|d| d.sort_key());
```

Comment:
I could be wrong here, but it seems that unstable sorting may break sprite batching, as `sort_key()` only contains the z coordinate but not the entity id.

Reply:
I see no reason for this to block other render phases from using a faster sort though. Added an associated constant on `PhaseItem` to demark whether it's safe to use an unstable sort or not, and added an if statement to branch on it. The compiler will optimize out the branch that isn't taken. 

We could probably also extend it to be an enum of `SortPhaseMode` as follows:

```rust
pub enum SortPhaseMode {
   /// Requires a stable sort. Generally required for proper batching based on external criteria.
   Stable,
   /// The default: allows unstable sorting. Usually faster than Stable.
   Unstable,
   /// Unsorted. Omits sorting entirely.
   None,
}
```

Reply:
Okay, go ahead with this PR then. The problems with sprites can be tackled later.

Reply:
This has been implemented separately as of #5049.